internal/monitors/metricsfilter.go

```diff
@@ -0,0 +1,144 @@
+package monitors
+
+// Filter of datapoints based on included status and user configuration of
+// extraMetrics and extraGroups.
+
+import (
+	"errors"
+	"fmt"
+	"github.com/signalfx/golib/datapoint"
+	"github.com/signalfx/signalfx-agent/internal/core/dpfilters"
+	"github.com/signalfx/signalfx-agent/internal/utils/filter"
+	"github.com/sirupsen/logrus"
+	"strings"
+)
+
+var _ dpfilters.DatapointFilter = &extraMetricsFilter{}
+
+type extraMetricsFilter struct {
+	metadata     *Metadata
+	extraMetrics map[string]bool
+	stringFilter *filter.BasicStringFilter
+}
+
+func validateMetricName(metadata *Metadata, metricName string) error {
+	if strings.TrimSpace(metricName) == "" {
+		return errors.New("metric name cannot be empty")
+	}
+
+	if !metadata.MetricsExhaustive {
+		// The metrics list isn't exhaustive so can't do extra validation.
+		return nil
+	}
+
+	if strings.ContainsRune(metricName, '*') {
+		// Make sure a metric pattern matches at least one metric.
+		m, err := filter.NewBasicStringFilter([]string{metricName})
+		if err != nil {
+			return err
+		}
+
+		for metric := range metadata.Metrics {
+			if m.Matches(metric) {
+				return nil
+			}
+		}
+
+		logrus.Warnf("extraMetrics: metric pattern '%s' did not match any available metrics", metricName)
+	}
+
+	if !metadata.HasMetric(metricName) {
+		logrus.Warnf("extraMetrics: metric '%s' does not exist", metricName)
+	}
+
+	return nil
+}
+
+func validateGroup(metadata *Metadata, group string) ([]string, error) {
+	if strings.TrimSpace(group) == "" {
+		return nil, errors.New("group cannot be empty")
+	}
+
+	metrics, ok := metadata.GroupMetricsMap[group]
+	if !ok {
+		return nil, fmt.Errorf("group %s does not exist", group)
+	}
+	return metrics, nil
+}
+
+func newMetricsFilter(metadata *Metadata, extraMetrics, extraGroups []string) (*extraMetricsFilter, error) {
+	var filterItems []string
+
+	for _, metric := range extraMetrics {
+		if err := validateMetricName(metadata, metric); err != nil {
+			return nil, err
```

Comment:
This bubbles up to preventing the monitor from starting, when it should really just be a warning as it isn't a critical error that should prevent monitoring.